source/extensions/common/wasm/wasm.cc

```diff
@@ -68,13 +68,13 @@ inline Wasm* getWasm(WasmHandleSharedPtr& base_wasm_handle) {
 
 void Wasm::initializeLifecycle(Server::ServerLifecycleNotifier& lifecycle_notifier) {
   auto weak = std::weak_ptr<Wasm>(std::static_pointer_cast<Wasm>(shared_from_this()));
-  lifecycle_notifier.registerCallback(Server::ServerLifecycleNotifier::Stage::ShutdownExit,
-                                      [this, weak](Event::PostCb post_cb) {
-                                        auto lock = weak.lock();
-                                        if (lock) { // See if we are still alive.
-                                          server_shutdown_post_cb_ = std::move(post_cb);
-                                        }
-                                      });
+  shutdown_handler_ = lifecycle_notifier.registerCallback(
```

Comment:
I don't really understand this code - can you explain briefly? Why is "this" held twice directly and via weak? Why is it posting the completion to its own dispatcher without doing any work?

Reply:
Taking a deeper look, I don't have a full history for why this server_shutdown_post_cb_ is needed. The previous shutdown logic is 100% no-op since it is never registered. Weak here is going to check if this object is still alive, and then the callback can be executed. The original implementation is posting the `completion_cb` in the destructor leading to issues which I think the root cause is: since it is latching the `completion_cb`, the [cb_guard](https://github.com/envoyproxy/envoy/blob/cadbc9b58e73d3a893a1ccb4cfca51d311c9e1db/source/server/server.cc#L1108-L1122) will never be executed, so that the server will not shutdown until it is timeout, and then further cause other segment issues.

My current version also has a issue to bypass some callbacks by posting `completion_cb` which is (dispatcher_.exit()) directly.  Based on the above analysis, I tend to remove the whole shutdown registration code block since it is never used, WDYT?

Reply:
+1, it seems like the only purpose is to make sure that the Wasm instance is deleted before proceeding with shutdown, but was no-op before. Removing it SGTM as well.